### DISCUSSION TITLE
Piccolo Admin not adding row in the table for complex schema.

### DISCUSSION THREAD

**Discussion by sumitsharansatsangi** [piccolo-orm/piccolo_admin#220] [General]
For complex schema, piccolo admin not adding rows. 
It is fetching the rows from the database perfectly, but the problem occurs while adding the row.
Pressing the `create` button is not adding the data to the database.

**Comment by sinisaos** [comment]
@sumitsharansatsangi Are there any errors or? Can you share the code for your database tables that we can try and see where the problem is?

**Reply by sumitsharansatsangi** [reply]
No, there is no any error. Sharing with you privately.

**Reply by sinisaos** [reply]
You can send it to the email that is on my Github account. I only need the content from `tables.py`, i.e. the table that causes problems (**not the data**) so that I can try it and try to understand what the problem is, I do not guarantee anything that I will succeed.

**Reply by sumitsharansatsangi** [reply]
@sinisaos , My table is as follows
```python
class TableA(Table):
    name= Varchar()

class TableB(Table):
    name=Varchar()
    accountNumber = BigInt()
    fk = ForeignKey(TableA)

class TableC(Table):
    name=Varchar()
    designation =Varchar()
    fk = ForeignKey(TableB)

class TableD(Table):
    name=Varchar()
    fk = ForeignKey(TableC)

class TableE(Table):
    name=Varchar()
    fk = ForeignKey(TableD)
```

**Reply by sumitsharansatsangi** [reply]
@dantownsend  has just resolved migration related issue for this schema, now I am facing admin related issue.

**Reply by sumitsharansatsangi** [reply]
It was also with the piccolo - 0.89.0 and also with 0.90.0.

**Reply by dantownsend** [reply]
I tried it locally, and can't replicate it at the moment.

Do you get any errors from the server - maybe a 500? Or anything in the Javascript console in the web browser?

Make sure you have the latest versions of `piccolo_admin` and `piccolo_api` too, as the latest versions show error messages a bit better in the UI.

**Reply by sumitsharansatsangi** [reply]
I have also tried it locally, what surprised me is that `create` button is not sending request to the server, neither giving me any error in the browser.

**Reply by sumitsharansatsangi** [reply]
I am using everything of the latest version.

**Reply by sumitsharansatsangi** [reply]
just inspected in the browser, received following error, when I pressed `Create` button.
```
vue.runtime.esm.js:1897 TypeError: Cannot read properties of undefined (reading 'type')
    at AddRowForm.vue?f4b9:45:56
    at u (runtime.js:63:40)
    at Generator._invoke (runtime.js:294:22)
    at Generator.next (runtime.js:119:21)
    at r (asyncToGenerator.js:3:20)
    at l (asyncToGenerator.js:25:9)
    at asyncToGenerator.js:32:7
    at new Promise (<anonymous>)
    at asyncToGenerator.js:21:12
    at a.submitForm (AddRowForm.vue?f4b9:33:31)
    ```
Pressing create button is not sending request to the server.
Pressing `create` button , prints `I was pressed and gives above error`

**Reply by sumitsharansatsangi** [reply]
For table that is not related like this, I am able to add row very nicely.

**Reply by dantownsend** [reply]
OK, that helps. The front end decides what to do based on the table 'schema'.

If you go to this URL, do you get any errors?

```
/api/tables/my-table/schema/
```

For example, if your table is called `movie`, and you're running the admin at `localhost:8000/admin`, go to this URL:

```
localhost:8000/admin/api/tables/movie/schema/
```

**Reply by sumitsharansatsangi** [reply]
![Screenshot at 2022-09-07 21-22-531](https://user-images.githubusercontent.com/45959281/188923801-6400a8b7-dcc3-4a2d-bbca-090c33273665.png)
Screenshot of the error in the browser.
@dantownsend , this is not the answer of your last suggestion. This is the error I am facing in the admin. Let me try what you have said.

**Reply by sumitsharansatsangi** [reply]
Surprisingly, I received, ` Auth failed`. In the other tab , I am getting data perfectly, even after getting this error in this ( http://localhost:8000/admin/api/tables/movie/schema/) table. (Not writing my original table name here)

**Reply by dantownsend** [reply]
OK, it sounds like it could be an authentication issue. Your app is trying to fetch the table's schema, and it can't for some reason.

Things to try off the top of my head:

* Log out and log in again
* Make sure that your Piccolo user has `active`, `admin`, and `superuser` set to `True`.

**Reply by sinisaos** [reply]
@sumitsharansatsangi Unfortunately I can't try it locally at the moment, but Piccolo Admin use session auth and you should be logged in from any browser tab. Try what Daniel said makes sense.

**Reply by sumitsharansatsangi** [reply]
Tried with ( http://127.0.0.1:8000/admin/api/tables/movie/schema/) gives the schema.
Changing from localhost to 127.0.0.1 works.

**Reply by sumitsharansatsangi** [reply]
Ignore authentication error, it was my fault that I have tried with localhost instead of 127.0.0.1. 
Please don't consider it as authentication issue, @dantownsend  and @sinisaos .

**Reply by dantownsend** [reply]
OK, cool.

The error in the front end makes it seem like either:

* The front end couldn't fetch the schema for some reason
* OR, the schema is somehow wrong. The error is saying that the `type` field shown below doesn't exist. Is that true in your schema?

<img width="622" alt="Screenshot 2022-09-07 at 17 17 16" src="https://user-images.githubusercontent.com/350976/188928959-138c1e47-146a-4f1a-975e-c3f01b67dc4e.png">

**Reply by sumitsharansatsangi** [reply]
I receive following with this url ( http://127.0.0.1:8000/admin/api/tables/tablename1/schema/)
Please see, if anything wrong in the schema, only changed column_name and tablename, rest is as it was.
```
{
  "title": "tablename1In",
  "type": "object",
  "properties": {
    "column_name1": {
      "title": "Column_name1",
      "extra": {
        "help_text": null,
        "choices": null
      },
      "nullable": false,
      "maxLength": 255,
      "type": "string"
    },
    "column_name2": {
      "title": "Column_name2",
      "extra": {
        "help_text": null,
        "choices": null
      },
      "nullable": false,
      "type": "integer"
    },
    "column_name3": {
      "title": "Column_name3",
      "extra": {
        "help_text": null,
        "choices": null
      },
      "nullable": false,
      "type": "integer"
    },
    "column_name4": {
      "title": "Column_name4",
      "extra": {
        "help_text": null,
        "choices": null
      },
      "nullable": false,
      "type": "integer"
    },
    "column_name5": {
      "title": "Column_name5",
      "extra": {
        "help_text": null,
        "choices": null
      },
      "nullable": false,
      "type": "integer"
    },
    "column_name6": {
      "title": "Column_name6",
      "extra": {
        "help_text": null,
        "choices": null
      },
      "nullable": false,
      "type": "integer"
    },
    "column_name7": {
      "title": "Column_name7",
      "extra": {
        "help_text": null,
        "choices": null
      },
      "nullable": false,
      "type": "integer"
    },
    "column_name8": {
      "title": "Column_name8",
      "extra": {
        "help_text": null,
        "choices": null
      },
      "nullable": true,
      "type": "array",
      "items": {
        "type": "string"
      }
    }
  },
  "help_text": null,
  "visible_column_names": [
    "id",
    "column_name1",
    "column_name2",
    "column_name3",
    "column_name4",
    "column_name5",
    "column_name6",
    "column_name7",
    "column_name8"
  ],
  "visible_filter_names": [
    "id",
    "column_name1",
    "column_name2",
    "column_name3",
    "column_name4",
    "column_name5",
    "column_name6",
    "column_name7",
    "column_name8"
  ],
  "rich_text_columns": [],
  "media_columns": [],
  "visible_fields_options": [
    "id",
    "column_name1",
    "column_name2",
    "column_name3",
    "column_name4",
    "column_name5",
    "column_name6",
    "column_name7",
    "column_name8"
  ],
  "primary_key_name": "id"
}
```

**Reply by dantownsend** [reply]
That looks fine. Hmm, are there any other errors in the Javascript console?

**Reply by sumitsharansatsangi** [reply]
@dantownsend , @sinisaos , shared my schema, please have a look.

**Reply by sumitsharansatsangi** [reply]
No, What I have shared was the only error.
![Screenshot at 2022-09-07 21-22-531](https://user-images.githubusercontent.com/45959281/188934083-d7313b95-0b85-4d84-a19f-95e376907388.png)

**Reply by sinisaos** [reply]
The schema looks good to me. All columns have a proper type (string, integer).

**Reply by sumitsharansatsangi** [reply]
Consider , almost similar type of schema (Some table have two or three more field , some has less ), and are related as follows
Table1 -> Table2 -> Table3 -> Table 4 -> Table->5.
Will this create any issue ?

**Reply by sumitsharansatsangi** [reply]
Because, My schema is almost like that. I have shared the schema of Table1(The Grand Grand Parent)

**Reply by dantownsend** [reply]
I don't think so - I tried it with this schema and it seemed to work:

```python
from piccolo.columns import Varchar, ForeignKey
from piccolo.table import Table


class Manager(Table):
    name = Varchar()

class Manager2(Table):
    fk = ForeignKey(Manager)

class Manager3(Table):
    fk = ForeignKey(Manager2)

class Manager4(Table):
    fk = ForeignKey(Manager3)

class Manager5(Table):
    fk = ForeignKey(Manager4)
```

There could be a bug somewhere, but I don't think it's related to the foreign keys in this case.

Have all of the database migrations been applied successfully? Might be worth checking the database to make sure it's what your expect.

**Reply by sumitsharansatsangi** [reply]
But, since create button is not sending request to my server , when I am adding the row, how can it be a database related issue, Also I have manually added row to the database and piccolo-admin fetched it correctly, without any issue. 

I'll request you try with schema with a lots of column and like I have. Also use Array as a column.

**Reply by sumitsharansatsangi** [reply]
Since it is displaying the data correctly, I don't think it is database or schema related issue. It can be something else.

**Reply by sumitsharansatsangi** [reply]
Hope, this will help you.
![Screenshot at 2022-09-07 22-30-531](https://user-images.githubusercontent.com/45959281/188937678-53c62acc-3a50-405f-b8bc-4434f0bdf80a.png)


![Screenshot at 2022-09-07 22-30-251](https://user-images.githubusercontent.com/45959281/188937699-6f43d5a7-58bc-4509-9e31-02431230eee9.png)

**Reply by sumitsharansatsangi** [reply]
Please consider my this statement. "Pressing the `create` button is not sending request to the server"

**Reply by dantownsend** [reply]
It might be something to do with the name of the column.

This is the problematic line:

```javascript
} else if (this.schema.properties[key].type == "array") {
```

We could do with putting a debug statement and seeing what the value of `key` is. It must not be matching what's in schema for some reason.

**Reply by dantownsend** [reply]
I need to go for a bit - will be back later.

**Reply by sumitsharansatsangi** [reply]
I have just observed, that I have more than seven column as big int or integer, out of which the second last column is not filled with zero by default in the admin panel , whereas rest are coming filled with zero, Why is it so?

**Reply by sumitsharansatsangi** [reply]
I think schema of that column is creating error.

**Reply by sumitsharansatsangi** [reply]
Hey @dantownsend , Can't you become a magician for a minute, you know spell, the error was because of the column name .
My column name was `columnName` changing it to `column_name` works.

**Reply by sumitsharansatsangi** [reply]
@dantownsend , @sinisaos , If this is a bug, please correct it, if it was intentionally, please add it in the documentation., If I have missed to read it, pardon me for wasting your time.

**Reply by sinisaos** [reply]
@sumitsharansatsangi I just managed to try your scheme. @dantownsend and you are right that the column name causes problem. Camelcase variable name is not common in Python and I think this [line](https://github.com/piccolo-orm/piccolo_admin/blob/42953dee8b753e49a1a7931b41d9797cca77d349/admin_ui/src/components/AddRowForm.vue#L39) caused the problem because it cannot split the column name properly (`accountNumber` became `Accountnumber` in Piccolo Admin schema title name and it can't be splitted).

**Reply by sumitsharansatsangi** [reply]
@sinisaos , it's true that Camelcase variable name is not common in python. I had just because of copying the schema from another language and just changing it to be compatible with piccolo, I was unware that using camelCase with an issue. I had similar experience earlier too,but that time I was  using Django, django raise some error while doing migrations of the table (or somewhere else, I don't remember exactly ), if similar can also be done for the piccolo, it will be helpful for other developers. @dantownsend , till raising exception is not added to the piccolo, please add in the schema section of the piccolo documentation to use snake_case as variable name.

Thank you both for the help.

**Reply by dantownsend** [reply]
Cool, I'm glad we know what caused the problem.

We assume that that column names are snake case. I think this can be fixed though in Piccolo Admin.

**Reply by sumitsharansatsangi** [reply]
If it can be fixed, please fix it.

**Reply by dantownsend** [reply]
This should be fixed now in `piccolo_admin==0.32.0`.
